2/answers.txt

```diff
@@ -0,0 +1,231 @@
+
+Q17: What is NGINX and why is it considered preferable to Apache?
+
+NGINX is a high performance load balancer and reverse proxy service. 
+The reason the architecture used in NGINX is considered better than Apache is because of the old mod_php module in Apache that would embed a php interpreter for every single process,
+even when the information being served was not PHP. Also, the way in which Apache creates a new process for every connection, severely limiting its ability to multithread and use multiple cores
+making a large number of processes become quickly an expensive activity.
+
+Whereas Nginx uses asynchronous, non-blocking, event driving architecture, allowing for a far more optimal handling of processes, ideally with one process per core, and every thread a subset thereafter.
+
+Q18: What is a page file? What happens to applications that exhaust all available system memory? Is 'paging' bad?
+
+Page files are like caches in that they store data that RAM hasnt used in a while, as a single contiguous chunk of data, so it can be read faster from the HD than from its original location.
+Applications that exhaust all system memory are probably experiencing a leak of some kind, as memory that is no longer required should be shadow dumped and that memory freed up. If however this does happen.
+Usually the program, or the system itself with entirely freeze up and crash, as critical processes are unable to find memory to function correctly.
+
+Paging isnt bad inherently, but poor paging management can be a crutch to a poor design, and when that crutch eventually fails, it will collapse everything.
+
+Q19: Why is dynamic memory, also known as memory ballooning, a useful technology for virtual machines?
+
+For virtual machines, having dynamic memory is important and useful because if a single VM is using less memory than another, memory can be re-allocated or reclaimed to be used by the other vM instances.
+This is particularly useful because it can lower costs than statically ensuring you have enough memory for max output of each VM individualy. but also if you are hosting a set of VMs
+Where the contiguous user or vm usage fluctuates, you can have a smaller stack of memory that can cycle between VMs if the overall max output of all VMs is not higher than the maximum memory available.
+
+Q20: What are processes? Can they share memory?
+
+Processes are instances of programs executing on a core, made up a single thread or multiple threads, there are many forms of IPC (Inter Process Communication) and one of them 
+is the use of shared memory segments, whereby mulitple processes can access the same memory segment. In Unix this is called anonymous shared memory. whereas windows has a native shared memory function.
+Multiple processes cannot share main memory space however, and the creation of share memory segments is a relatively resource expensive task, and should only be used if it would create a larger benefit.
```

Comment:
Sharing memory between processes (at scale) is usually achieved by using a service running on a separate process (or even separate server if inside a data centre) that implements a key/value in-memory store, or similar.

An example of such a service is Redis.

For smaller projects where mass horizontal scalability is not required and everything can be run on a single machine with multiple cores, IPC is used to communicate between processes. This has a performance overhead as messages passed between processes have to be serialised when sent and de-serialised on receipt.

Serialisation is a non-trivial problem for computers. Moving large datasets between processes is far less efficient than using shared memory via threading as no serialisation needs to occur.

Reply:
Note that serialisation is inherently a deep clone of the data. nothing is shared between the processes